### DISCUSSION TITLE
[Change Log] Canokey Pigeon/Canary 固件版本更新说明

### DISCUSSION THREAD

**Discussion by z4yx** [canokeys/canokey-product#20] [Announcements]
## 3.0.3 -> NEXT
- OpenPGP: 支持SECP521R1算法

## 3.0.2 -> 3.0.3 (2025.3)
- OpenPGP: 修正Algorithm Info DO的编码格式（暂无应用使用该DO）
- PIV: 解决数据长度不等于32时ED25519算法产生签名错误的问题

## 3.0.0 -> 3.0.2 (2025.2)
- 允许读取NFC使能状态（不需要验证admin PIN）
- U2F: 修复前一版本触摸不可用的问题
- FIDO2: 避免CCID和HID协议同时访问时出错，解决在最新macOS不可用的问题
- PIV: 修复前一版本25519密钥导入的问题
- OATH: 修复分多次返回响应时，状态码不正确的问题（该问题不影响 CanoKey Console 的使用）
- 移除未使用的Meta应用

## 2.0.1 -> 3.0.0 (2024.11)
- 增加CCID和WebUSB访问冲突解决机制
- 增加NFC功能开关
- Pass: 新增静态密码存储和输入功能，触摸时模拟键盘输入密码
- OpenPGP: 支持GET CHALLENGE命令，可用作随机数发生器
- PIV: 扩展算法支持，RSA3072/4096，ED25519，SECP256K1
- PIV: 新增DO两个，Fingerprints和Facial Images，可存储自定义数据

## 2.0.0 -> 2.0.1 (2023.11)
- FIDO: 解决新版Firefox上注册失败的问题。具体来说，当extension map为空时不返回map，避免Firefox使用的库中出现反序列化错误（mozilla/authenticator-rs#317）。

## 1.6.2 -> 2.0.0 (2023.11)
- FIDO: 更新到 FIDO 2.1 版本，支持 Credential Management, Credential Protection, Credential Blob, Large Blob Key 特性，修正密钥生成bug (#19)
- PIV: 支持配置密钥的属性（是否要求触摸，是否要求验证PIN），增加 82/83 两组密钥，支持一些元数据以增强兼容性
- OpenPGP: 支持 RSA3072/4096 算法，修正上电后首次ED25519签名错误问题
- OATH: 支持返回完整 hash 结果

## 1.6.1 -> 1.6.2 (2022.10)
- RSA: 支持 e != 65537 的公钥导入
- OATH: 可配置触摸输出 HOTP code 时是否加回车

## 1.5.2 -> 1.6.1 (2022.06)
- Driver: 修正 USB 不能正常复位的问题（影响在虚拟机中使用），修正 NFC 射频芯片EEPROM配置问题
- PIV: 支持超过 3K 长度的证书，修正自动生成的 CCC/CHUID 中的格式错误

## 1.5.2 (2021.12)
Canokey Pigeon 首个正式版本

**Comment by Mino9526** [comment]
已售出的令牌 支持固件更新吗

**Reply by dangfan** [reply]
不支持

**Reply by MetropolisRaccoon** [reply]
Will there be any upcoming plans for users to flash and upgrade firmwares?
